ibis/backends/dask/core.py

```diff
@@ -0,0 +1,544 @@
+:class:`~ibis.expr.operations.ScalarParameter`, in which case the ``scope``
+passed to ``post_execute`` would be the bound values passed in at the time the
+``execute`` method was called.
+
+
+Scope
+-------------------
+Scope is used across the execution phases, it iss a map that maps Ibis
+operators to actual data. It is used to cache data for calculated ops. It is
+an optimization to reused executed results.
+
+With time context included, the key is op associated with each expression;
+And scope value is another key-value map:
+- value: pd.DataFrame or pd.Series that is the result of executing key op
+- timecontext: of type TimeContext, the time context associated with the data
+stored in value
+
+See ibis.common.scope for details about the implementaion.
+"""
+
+from __future__ import absolute_import
+
+import datetime
+import functools
+import numbers
+from typing import Optional
+
+import dask.dataframe as dd
+import numpy as np
+from multipledispatch import Dispatcher
+from pandas import Timedelta, Timestamp
+
+import ibis
+import ibis.common.exceptions as com
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+import ibis.expr.types as ir
+import ibis.expr.window as win
+from ibis.client import find_backends
+from ibis.expr.scope import Scope
+from ibis.expr.timecontext import canonicalize_context
+from ibis.expr.typing import TimeContext
+
+from .dispatch import execute_literal, execute_node, post_execute, pre_execute
+from .trace import trace
+
+integer_types = np.integer, int
+floating_types = (numbers.Real,)
+numeric_types = integer_types + floating_types
+boolean_types = bool, np.bool_
+fixed_width_types = numeric_types + boolean_types
+date_types = (datetime.date,)
+time_types = (datetime.time,)
+timestamp_types = Timestamp, datetime.datetime, np.datetime64
+timedelta_types = Timedelta, datetime.timedelta, np.timedelta64
+temporal_types = date_types + time_types + timestamp_types + timedelta_types
+scalar_types = fixed_width_types + temporal_types + (dd.core.Scalar,)
+simple_types = scalar_types + (str, type(None))
+
+
+@functools.singledispatch
+def is_computable_input(arg):
+    """All inputs are not computable without a specific override."""
+    return False
+
+
+@is_computable_input.register(ibis.client.Client)
+@is_computable_input.register(ir.Expr)
+@is_computable_input.register(dt.DataType)
+@is_computable_input.register(type(None))
+@is_computable_input.register(win.Window)
+@is_computable_input.register(tuple)
+def is_computable_input_arg(arg):
+    """Return whether `arg` is a valid computable argument."""
+    return True
+
+
+# Register is_computable_input for each scalar type (int, float, date, etc).
+# We use consume here to avoid leaking the iteration variable into the module.
+ibis.util.consume(
+    is_computable_input.register(t)(is_computable_input_arg)
+    for t in scalar_types
+)
+
+
+def execute_with_scope(
+    expr,
+    scope: Scope,
+    timecontext: Optional[TimeContext] = None,
```

Comment:
make a todo item to combine a number of these execution primitives for pandas/dask

Reply:
though you *might* simply be able to import the pandas ones directly

Reply:
yeah definitely try to do this here (if it gets bogged down, then just create a todo)

Reply:
Yeah - you're right - not sure why I didn't import these previously.